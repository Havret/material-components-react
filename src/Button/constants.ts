const BUTTON = 'mdc-button';
const BUTTON_ICON = 'mdc-button__icon';
const BUTTON_RAISED = 'mdc-button--raised';
const BUTTON_UNELEVATED = 'mdc-button--unelevated';
const BUTTON_STROKED = 'mdc-button--stroked';
const BUTTON_DENSE = 'mdc-button--dense';
const BUTTON_COMPACT = 'mdc-button--compact';

export { BUTTON, BUTTON_ICON, BUTTON_RAISED, BUTTON_UNELEVATED, BUTTON_STROKED, BUTTON_DENSE, BUTTON_COMPACT }