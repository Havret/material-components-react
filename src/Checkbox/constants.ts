export const ROOT = 'mdc-checkbox';
export const UPGRADED = 'mdc-checkbox--upgraded';
export const NATIVE_CONTROL = 'mdc-checkbox__native-control';
export const DISABLED = 'mdc-checkbox--disabled';
export const BACKGROUND = 'mdc-checkbox__background';
export const CHECKMARK = 'mdc-checkbox__checkmark';
export const CHECKMARK_PATH = 'mdc-checkbox__checkmark__path';
export const MIXEDMARK = 'mdc-checkbox__mixedmark';

export const ANIM_UNCHECKED_CHECKED = 'mdc-checkbox--anim-unchecked-checked';
export const ANIM_UNCHECKED_INDETERMINATE = 'mdc-checkbox--anim-unchecked-indeterminate';
export const ANIM_CHECKED_UNCHECKED = 'mdc-checkbox--anim-checked-unchecked';
export const ANIM_CHECKED_INDETERMINATE = 'mdc-checkbox--anim-checked-indeterminate';
export const ANIM_INDETERMINATE_CHECKED = 'mdc-checkbox--anim-indeterminate-checked';
export const ANIM_INDETERMINATE_UNCHECKED = 'mdc-checkbox--anim-indeterminate-unchecked';

export const ANIM_END_LATCH_MS = 100;

export enum CheckState {
    checked,
    unchecked,
    indeterminate
}