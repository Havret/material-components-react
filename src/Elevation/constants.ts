export const ELEVATION = 'mdc-elevation--z';
export const ELEVATION_TRANSITION = 'mdc-elevation-transition';