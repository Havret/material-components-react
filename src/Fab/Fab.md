A floating action button component

### FABs with Ripple

```js
<div className="margin">
    <Fab icon="favorite"/>
    <Fab mini icon="favorite"/>                      
</div>
```

### CSS Only FABs

```js
<div className="margin">
    <Fab ripple={false} icon="favorite"/>
    <Fab ripple={false} mini icon="favorite"/>                      
</div>
```

### Example of Enter and Exit Motions

```js
initialState = { exited: false };
<div className="margin">
<Fab exited={state.exited} icon="favorite"/>
<Button stroked onClick={() => setState(prev => ({exited: !prev.exited}))}>Toggle Exited</Button>
</div>
```