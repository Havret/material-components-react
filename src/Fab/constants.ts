export const FAB = 'mdc-fab';
export const FAB_ICON = 'mdc-fab__icon';
export const FAB_MINI = 'mdc-fab--mini';
export const FAB_EXITED = 'mdc-fab--exited';