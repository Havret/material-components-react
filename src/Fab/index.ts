export { Fab as default, Fab, FabProps } from './Fab';
