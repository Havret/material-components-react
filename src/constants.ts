export const MATERIAL_ICONS = 'material-icons';