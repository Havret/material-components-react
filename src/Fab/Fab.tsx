import * as cx from 'classnames';
import * as React from 'react';
import { ObjectOmit } from '../utils';
import { HTMLProps, MouseEventHandler } from 'react';
import { FAB, FAB_EXITED, FAB_ICON, FAB_MINI } from './constants';
import { MATERIAL_ICONS } from '../constants';
import { Ripple, RippleComponentProps } from '../Ripple';

interface FabProps extends ObjectOmit<HTMLProps<HTMLButtonElement>, 'disabled'> {
    /**
     * Addition CSS classes that will be added to the root element.
     */
    className?: string;
    /**
     * The name of the icon font ligature.
     */
    icon?: string;
    /**
     * If `true`, modifies the FAB to a smaller size.
     */
    mini?: boolean;
    /**
     * Animates the FAB out of view. When this class is removed, the FAB will return to view.
     */
    exited?: boolean;
    /**
     * If `true` adds a ripple effect to the component
     */
    ripple?: boolean;
    /**
     * Click event handler.
     */
    onClick?: MouseEventHandler<HTMLButtonElement>;
}

class Fab extends React.Component<FabProps, {}> {
    static defaultProps = {
        mini: false,
        exited: false,
        ripple: true
    };

    render() {
        return this.props.ripple
            ? <Ripple surface={false} render={this.renderFab}/>
            : this.renderFab();
    }

    private renderFab = (rippleProps: Partial<RippleComponentProps> = {}) => {
        const {className: rippleClass, ...rippleRest} = rippleProps;

        const {icon, mini, exited, className, ripple, ...rest} = this.props;

        const classNames = cx(FAB, MATERIAL_ICONS, rippleClass, {
            [FAB_MINI]: mini,
            [FAB_EXITED]: exited
        }, className);

        return (
            <button {...rest} {...rippleRest} className={classNames}>
                {icon && <span className={FAB_ICON}>{icon}</span>}
            </button>
        );
    };
}

export { Fab as default, Fab, FabProps };