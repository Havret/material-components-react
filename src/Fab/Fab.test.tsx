import { configure, shallow, mount } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import * as React from 'react';
import toJson from 'enzyme-to-json';

import { cssClasses as rippleCssClasses } from '../Ripple/constants';
import { Fab } from './Fab';
import { FAB, FAB_EXITED, FAB_MINI } from './constants';

describe('Fab', () => {
    configure({adapter: new Adapter()});

    test('It should have mdc classes - Ripple Disabled', () => {
        const wrapper = shallow(<Fab ripple={false}/>);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('It should have mdc classes - Ripple Enabled', () => {
        const wrapper = mount(<Fab/>);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('It should add extra class to the root element', () => {
        const extraClass = 'extra-class';
        const wrapper = shallow(<Fab className={extraClass} ripple={false}/>);
        expect(wrapper.hasClass(extraClass)).toBeTruthy();
    });

    test('It should not be mini', () => {
        const wrapper = shallow(<Fab ripple={false}/>);
        expect(wrapper.hasClass(FAB_MINI)).toBeFalsy();
    });

    test('It should be mini', () => {
        const wrapper = shallow(<Fab ripple={false} mini={true}/>);
        expect(wrapper.hasClass(FAB_MINI)).toBeTruthy();
    });

    test('It should not be exited', () => {
        const wrapper = shallow(<Fab ripple={false} />);
        expect(wrapper.hasClass(FAB_EXITED)).toBeFalsy();
    });

    test('It should be exited', () => {
        const wrapper = shallow(<Fab ripple={false} exited={true} />);
        expect(wrapper.hasClass(FAB_EXITED)).toBeTruthy();
    });

    test('It should be ripple upgraded when when ripple enabled', () => {
        const wrapper = mount(<Fab ripple={true}/>);
        expect(wrapper.find(`.${FAB}`).hasClass(rippleCssClasses.ROOT)).toBeTruthy();
    });
});