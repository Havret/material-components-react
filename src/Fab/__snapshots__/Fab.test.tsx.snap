// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fab It should have mdc classes - Ripple Disabled 1`] = `
<button
  className="mdc-fab material-icons"
/>
`;

exports[`Fab It should have mdc classes - Ripple Enabled 1`] = `
<Fab
  exited={false}
  mini={false}
  ripple={true}
>
  <Ripple
    disabled={false}
    primary={false}
    render={[Function]}
    secondary={false}
    surface={false}
    unbounded={false}
  >
    <button
      className="mdc-fab material-icons mdc-ripple-upgraded"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "--mdc-ripple-fg-scale": "Infinity",
          "--mdc-ripple-fg-size": "0px",
          "--mdc-ripple-left": "0px",
          "--mdc-ripple-top": "0px",
        }
      }
    />
  </Ripple>
</Fab>
`;
